### ISSUE TITLE
Checking for versions of specific components

### ISSUE THREAD

**Issue by emodric** [symfony/symfony#35363]
Before Symfony 4, it was easy to check for installed version of Symfony by utilizing constants available in `Symfony\Component\HttpKernel\Kernel`.

However, with Symfony 4, now that the best practice is installing Symfony components individually, it can happen (due to requirements and conflicts in Composer), that some components are installed in different minor versions than the other components (or at least different from `HttpKernel` component).

This makes it practically impossible to check for the installed Symfony version at runtime, since now component versions are decoupled.

Thus, the following code, which works around the fact that `canonicalize` option was added in Symfony 4.1, would not work, if minor versions of `symfony/http-kernel` and `symfony/validator` were different:

```
$localeConstraint = Kernel::VERSION_ID >= 40100 ?
    new Constraints\Locale(['canonicalize' => true]) :
    new Constraints\Locale();
```

It would be great to add the version info to other components. Presumably, these versions are managed with a tool and not updated manually, so hopefully, it wouldn't bring too much complexity in the release process of Symfony.

**Comment by derrabus** [comment]
What problem are you trying to solve?

If you really need to query the exact version of a certain Symfony component at runtime, you can use `ocramius/package-versions` for this purpose. But I don’t think that we should add version constants to all Symfony packages.

**Comment by emodric** [comment]
The problem I'm trying to solve is that my product needs to support multiple versions of Symfony: 3.4, 4.4 and 5.0, and I can do that only in runtime with checks like these.

`ocramius/package-versions` cannot be used reliably, because if dev versions of Symfony packages are installed (e.g. for testing purposes), it yields commit hashes instead of specific versions.

**Comment by derrabus** [comment]
And feature detection would not work for you?

**Comment by nicolas-grekas** [comment]
We have the same need in core when writing forward/backward compatibility layers, and we've always found a way to do without version constants. There is always a new method/class to check for.

Please consider doing the same as we'd really like to not add these constants in components. Even with automation, that'd be added complexity we'd better avoid.

**Comment by emodric** [comment]
In most cases it can work, yes, but there are cases where it cannot, or where I'd need to use reflection, which I'd like to avoid.

Case in point: `Symfony\Component\HttpFoundation\Request::isMethodSafe` does not have an argument either in Symfony 3.4, 4.4 or 5.0. However, it throws the deprecation in Symfony 3.4. One of the ways I've found to check the version here would be `class_exists('Symfony\Component\HttpFoundation\Session\Storage\Handler\SessionHandlerFactory')`, but that sounds convoluted and not really related to code in question.

What I'm saying is, feature detection works, yes, but not always, and sometimes it can bring a couple WTFs along the way to future readers of the code.

**Comment by ro0NL** [comment]
i agree feature detection looks a bit clumsy sometimes ... but im not sure we can expect vendors to solve this issue, and expose their versions. Symfony is no exception here IMHO.

I'd expect some tool to build `composer_package_version_compare('sf/some', '3.4', '<=')`

**Comment by ro0NL** [comment]
> Case in point: Symfony\Component\HttpFoundation\Request::isMethodSafe does not have an argument either in Symfony 3.4, 4.4 or 5.0 (commented out in 3.4, non-existant in 4.4 or 5.0). However, it throws the deprecation in Symfony 3.4 when the argument is used with certain values.

shouldnt you fix the argument list as of 3.4?

**Comment by emodric** [comment]
> shouldnt you fix the argument list as of 3.4?

I was calling the method with `false` as the argument, to disable the deprecation, which was valid for 3.4. But anyways, this was just the first example I found in my code base to demonstrate the issue :) Surely, there are others like this :)

**Comment by emodric** [comment]
> `composer_package_version_compare('sf/some', '3.4', '<=')`

How would this work with dev versions of packages? Composer, and by extension `ocramius/package-versions` does not provide a semantic string of a version number for dev packages.

**Comment by ro0NL** [comment]
I see, 3.4 requires the argument actually. You could check for https://github.com/symfony/symfony/blob/b493997b71acc3bda554d3bce148d0abdd54395a/src/Symfony/Component/HttpFoundation/Request.php#L1578 using docblock reflection.

> How would this work with dev versions of packages?

it could take branch-alias into account, for commit hashes not sure what's possible without any GIT info. 

> components are not installed by themselves, instead symfony/symfony is used.

all component would be the symfony/symfony version then (which it replaces), so this can be resolved id say.

alternatively, deprecations can be taken for granted and deps can be bumped :)

TLDR; it's not a symfony issue IMHO.

**Comment by nicolas-grekas** [comment]
> ocramius/package-versions does not provide a semantic string of a version number for dev packages

Composer is aware of branch-aliases defined in composer.json files - I think you have a feature request to send to ocramius/package-versions :)

> this was just the first example I found in my code base to demonstrate the issue

If we take the examples one by one and all have a solution already, it means there are no issues. Don't try to overgeneralize, please :) Especially because the "solution" you ask for has new drawbacks - significant maintenance overhead - and nobody but Fabien will pay his free time on this. Better not. Unless again there is an example with no solution. But along the year, I've always found a solution to all similar situations - which means there are no generic issues, but only ad-hoc solutions - which are fine.

**Comment by emodric** [comment]
> alternatively, deprecations can be taken for granted and deps can be bumped :)

Easier said than done. We do have clients that pay for the product and expect it to work on their projects. I can't just go and out of the blue bump requirements.

> If we take the examples one by one and all have a solution already, it means there are no issues. Don't try to overgeneralize, please :)

I cannot know all examples where these version constants might be helpful, but I think it is a good way to consistently solve issues with supporting multiple versions. Other solutions offered here either require new dependencies (docblock reflection, ocramius/package-versions), are not reliable (what happens if later versions change conditions I'm using to detect a feature, without impacting the feature itself?) or can be convoluted.

Anyways, I see that you have made up your mind so I will end the discussion here.

Yes, I am aware that it will raise maintenance complexity, but IMO, it would be worth it.

**Comment by nicolas-grekas** [comment]
> (what happens if later versions change conditions I'm using to detect a feature, without impacting the feature itself?)

That doesn't happen: classes/methods/arguments don't disappear without a major version bump.
Again, from my practice of many years with the current alternatives, I've yet to see an issue with feature-detection. Convoluted? Sometimes I agree, but most of the time it's not that convoluted.

Thanks for opening the discussion, maybe the future will show you were right - but as of now, I'm personally resisting adding these constants - the cost/benefit balance is for the status quo IMHO, as you understood.

**Comment by xabbuh** [comment]
I agree with Nicolas about the cost of maintenance not being worth it. But I'd leave the final decision to @fabien as he is the one doing releases and so would be the one having to take the extra workload the suggested feature would require.

**Comment by Aerendir** [comment]
If I express an opinion, here I see two real truths.

I see at least two solutions alternative to the version constant:

1. Start linking this issue to all the others related to version identification, so we can start have a better idea of how spread is the problem;
2. Create a ne component (Symfony version checker) that does exactly this: support this kind of component directly by the Sf core team is way easier than reinventing the wheel each time each one on our own.

**Comment by Aerendir** [comment]
Also because finding each time a way to understand the version is time consuming: we need to read the code, the change log, test the solution... if we need to support three versione (3.4, 4 and 5) it is a lot of time... if we depend on three components, it is at least half a working day only to check the versions... 😅

**Comment by Aerendir** [comment]
I was thinking at this issue and at a certain point I reflected on the fact that each symfony package has a `CHANGELOG` file.
This file, obviously, has the last version of the component.

So, basically, if we read the file we can know the version of every component.

Obviously reading and parsing the file each time is a time consuming task. But the result can be cached.

This may be a good route: there is nothing to maintain other than the code that reads the `CHANGELOG` and caches the result.

Maybe this feature can be implemented in Flex, or in another component, or in another library...

**Comment by Aerendir** [comment]
> Additional problem with `ocramius/package-versions` is that if the my product is installed on Symfony 3.4, which is perfectly valid even if the product itself requires specific Symfony components, checks for the version of Symfony components will fail, since they are not installed, the main `symfony/symfony` package is.

Source: https://github.com/symfony/symfony/issues/35363#issuecomment-575175571

I was thinking at this: if the `symfony/symfony` package is installed, then you are sure the version is `3.4`. In other cases you can get the version using `ocramius/package-versions`, can't you?

Thinking at this issue, I reflected on the fact that each Symfony package has a `CHANGELOG` file that contains the last version of the installed package... It is another source of truth, also if reading and parsing the changelog may be a too much expensive task... But the result can also be cached...

**Comment by derrabus** [comment]
> I was thinking at this: if the `symfony/symfony` package is installed, then you are sure the version is `3.4`.

Actually, you _can_ still use the `symfony/symfony` monorepo package with Symfony 4 and 5 – it‘s just not recommended to do so. But in the monorepo scenario, you can be sure, that all components are installed in the version indicated by the constants on the `Kernel` class.

**Comment by javiereguiluz** [comment]
Let's close this issue because, at least for now, we only have the needed resources to keep doing the current feature detection. @emodric thanks for understanding and thanks for sending proposals to improve Symfony. We truly appreciate that!
